### PR TITLE
AsyncLocal diagnostics, clean slate

### PR DESCRIPTION
**Pull request by majocha**
We used `node` CE, a customized `async` CE variant to ensure `DiagnosticsThreadStatics` flow from thread to thread along with the execution of asynchronous compiler tasks.

`AsyncLocal<'T>` achieves this without the need for a specialized CE.

[from documetation:](https://learn.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?view=net-8.0)
> `AsyncLocal<T>` Class
> Represents ambient data that is local to a given asynchronous control flow, such as an asynchronous method.

This allows us to just remove `NodeCode` and replace `node` expressions with `async`.

Tests show `AsyncLocal` works as expected in scenarios involving `task`, `async`, `Async.SwitchToNewThread()` `SwitchToThreadPool()`, `Thread.Start()` as well as TPL parallel execution.

- [x] Added concurrent diagnostics logging for running many computations:

* `MultipleDiagnosticsLoggers.Parallel` generally for use  in place of `NodeCode.Parallel`. Replays diagnostics onto caller's logger preserving the given order of computations.
* `MultipleDiagnosticsLoggers.Sequential` replaces `NodeCode.Parallel`. Unlike `Async.Sequential` it starts immediately on the same thread so It automatically uses caller's diagnostics context.
   
- [x]  Added tests for `AsyncLocal` diagnostics flow in async and sync contexts.

Previous attempts: #16602, #16645

### CODE REVIEW

**Comment by majocha** [comment]
Might shed some light on what's going on here with the deadlocks:
https://github.com/dotnet/fsharp/pull/12638

Also:
https://devblogs.microsoft.com/premier-developer/the-danger-of-taskcompletionsourcet-class/

And maybe informative, since both GraphNode and AsyncMemoize that we have are variations on AsyncLazy:

https://github.com/dotnet/roslyn/blob/01b7c233fdda946c1a5628d4692ed827a07e33dd/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy%601.cs#L525-L530

        // We want to always run continuations asynchronously. Running them synchronously could result in deadlocks:
        // if we're looping through a bunch of Requests and completing them one by one, and the continuation for the
        // first Request was then blocking waiting for a later Request, we would hang. It also could cause performance
        // issues. If the first request then consumes a lot of CPU time, we're not letting other Requests complete that
        // could use another CPU core at the same time.
        public Request() : base(TaskCreationOptions.RunContinuationsAsynchronously)

**Comment by T-Gro** [comment]
Still a draft, or ready for review?

**Comment by majocha** [comment]
This was kinda put on hold as a risky change IIRC.
I can try to revive it but it would require a really careful look.

**Review by T-Gro — APPROVED** [review]

**Comment by majocha** [comment]
I added one more test to make sure it works with ListParallel. (In use with graph-based type checking).

File: src/Compiler/Driver/CompilerImports.fs [hunk 1]
```diff
@@ -2229,21 +2229,25 @@ and [<Sealed>] TcImports
 
     // NOTE: When used in the Language Service this can cause the transitive checking of projects. Hence it must be cancellable.
     member tcImports.RegisterAndImportReferencedAssemblies(ctok, nms: AssemblyResolution list) =
-        node {
+        async {
             CheckDisposed()
 
+            // Prevent deadlocks in FSI.
```
Comment (majocha) [inline src/Compiler/Driver/CompilerImports.fs:2235]: I'd love to put a more meaningful comment here, but I'm not 100% sure what's going on here and I don't want to misrepresent anything.

File: src/Compiler/Facilities/BuildGraph.fs [hunk 2]
```diff
@@ -5,217 +5,16 @@ module FSharp.Compiler.BuildGraph
...
-open FSharp.Compiler.DiagnosticsLogger
-open Internal.Utilities.Library
 
-[<NoEquality; NoComparison>]
-type NodeCode<'T> = Node of Async<'T>
+type Async with
 
-let wrapThreadStaticInfo computation =
-    async {
-        let diagnosticsLogger = DiagnosticsThreadStatics.DiagnosticsLogger
```
Comment (majocha) [inline src/Compiler/Facilities/BuildGraph.fs:18]: NodeCode does this stacking and unwinding implicitly. Without it, we need to be explicit about it by using functions like `CompilationGlobalsScope`, `UseDiagnistcsLogger` etc. Most compiler tasks do use them anyway, so there is little need to alter the existing code.

File: src/Compiler/Facilities/BuildGraph.fs [hunk 3]
```diff
@@ -274,36 +73,39 @@ type GraphNode<'T> private (computation: NodeCode<'T>, cachedResult: ValueOption
                                      ||| TaskContinuationOptions.NotOnFaulted
                                      ||| TaskContinuationOptions.ExecuteSynchronously)
                                 )
-                            |> NodeCode.AwaitTask
+                            |> Async.AwaitTask
+
+                        // Prevent deadlocks.
```
Comment (majocha) [inline src/Compiler/Facilities/BuildGraph.fs:78]: Again, this is something I'd need help with. I understand that the idea is to run this synchronously whenever possible to get better callstacks, but I've observed this often deadlocks in CI tests.

File: src/Compiler/Facilities/DiagnosticsLogger.fs [hunk 4]
```diff
@@ -905,16 +894,71 @@ type StackGuard(maxDepth: int, name: string) =
     static member GetDepthOption(name: string) =
         GetEnvInteger ("FSHARP_" + name + "StackGuardDepth") StackGuard.DefaultDepth
 
-type CaptureDiagnosticsConcurrently() =
-    let target = DiagnosticsThreadStatics.DiagnosticsLogger
-    let loggers = ResizeArray()
+type CaptureDiagnosticsConcurrently<'T>(computations: Async<'T> seq, ?eagerFormat) =
+    // Common error count for all computations.
+    let mutable errorCount = 0
```
Comment (majocha) [inline src/Compiler/Facilities/DiagnosticsLogger.fs:899]: This is not really useful in current implementation.

File: src/Compiler/Interactive/fsi.fs [hunk 5]
```diff
@@ -4618,7 +4617,7 @@ type FsiEvaluationSession
                 unresolvedReferences,
                 fsiOptions.DependencyProvider
             )
-            |> NodeCode.RunImmediateWithoutCancellation
+            |> Async.RunSynchronously
```
Comment (majocha) [inline src/Compiler/Interactive/fsi.fs:4620]: I can't recall why I changed this to RunSynchronously, probably when fighting with fsi deadlocks.

File: src/Compiler/Facilities/DiagnosticsLogger.fs [hunk 6]
```diff
@@ -905,16 +894,71 @@ type StackGuard(maxDepth: int, name: string) =
...
+            try
+                // We want to restore the current diagnostics context when finished.
+                use _ = new CompilationGlobalsScope()
+                return! forks.Computations |> method
+            finally
+                forks.ReplayDiagnostics.Wait()
+        }
+
+    let Parallel computations = computations |> run Async.Parallel
+    let Sequential computations = computations |> run Async.Sequential
```
Comment (majocha) [inline src/Compiler/Facilities/DiagnosticsLogger.fs:964]: Async.Sequential still uses thread pool, IIUC. Do we want a custom implementation instead?

Comment (T-Gro) [reply src/Compiler/Facilities/DiagnosticsLogger.fs:964]: It would make sense for `Sequential` to keep things running on the same thread, without using the ThreadPool.

File: src/Compiler/Facilities/BuildGraph.fs (hunk 2 above)
Comment (T-Gro) [reply src/Compiler/Facilities/BuildGraph.fs:18]: I thought this was the reason why the `AsyncLocal<..>` replacement was used, is it not?

File: src/Compiler/Facilities/BuildGraph.fs (hunk 3 above)
Comment (T-Gro) [reply src/Compiler/Facilities/BuildGraph.fs:78]: It deadlocked on the semaphore above, or somewhere else?

File: src/Compiler/Facilities/BuildGraph.fs (hunk 2 above)
Comment (majocha) [reply src/Compiler/Facilities/BuildGraph.fs:18]: Yes, what I meant was: on the boundary, for example `Async.AwaitNodeCode` does this wrapping implicitly.

OTOH AsyncLocal context will always pass across continuations.

File: src/Compiler/Facilities/DiagnosticsLogger.fs (hunk 6 above)
Comment (majocha) [reply src/Compiler/Facilities/DiagnosticsLogger.fs:964]: Yes, or at least start on the same thread.

File: src/Compiler/Facilities/BuildGraph.fs (hunk 2 above)
Comment (majocha) [reply src/Compiler/Facilities/BuildGraph.fs:18]: To clarify further, there are two orthogonal things here, as I understand it:
* Passing loggers between threads in the flow of an async computation. It's what `NodeCode` / `AsyncLocal` does.

 * Stacking / unwinding of diagnostics context. This is not really related to async, single threaded compiler tasks can do it. This is generally done explicitly with `UseDiagnosticsLogger` and such, but incidentally `NodeCode` does this in `AwaitNodeCode`, `AwaitAsync`, so it's possible we rely on this behavior somewhere.

File: src/Compiler/Facilities/BuildGraph.fs (hunk 3 above)
Comment (majocha) [reply src/Compiler/Facilities/BuildGraph.fs:78]: It froze the CI tests IIRC, I don't think I ever got to repro this one locally, so this is kind of a black box to me.

File: src/Compiler/Driver/CompilerImports.fs (hunk 1 above)
Comment (majocha) [reply src/Compiler/Driver/CompilerImports.fs:2235]: I can't reproduce the problem that this was supposed to solve. I'll get rid of this and see what happens in CI.

File: src/Compiler/Facilities/BuildGraph.fs (hunk 3 above)
Comment (majocha) [reply src/Compiler/Facilities/BuildGraph.fs]: I removed this and I don't see anything wrong. Oh well.

**Comment by majocha** [comment]
I think this is good for a review now.

File: src/Compiler/Utilities/illib.fs [hunk 7]
```diff
@@ -162,12 +162,17 @@ module internal PervasiveAutoOpens =
...
             let cancellationToken = defaultArg cancellationToken Async.DefaultCancellationToken
+
             let ts = TaskCompletionSource<'T>()
+
             let task = ts.Task
 
             Async.StartWithContinuations(computation, (ts.SetResult), (ts.SetException), (fun _ -> ts.SetCanceled()), cancellationToken)
 
-            task.Result
+            try
```
Comment (psfinaki) [inline src/Compiler/Utilities/illib.fs:172]: Why is this added?

File: src/Compiler/Driver/GraphChecking/GraphProcessing.fs [hunk 8]
```diff
@@ -180,7 +180,8 @@ let processGraphAsync<'Item, 'Result when 'Item: equality and 'Item: comparison>
         let! parentCt = Async.CancellationToken
         use localCts = new CancellationTokenSource()
 
-        let completionSignal = TaskCompletionSource()
+        let completionSignal =
+            TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously)
```
Comment (psfinaki) [inline src/Compiler/Driver/GraphChecking/GraphProcessing.fs:184]: Why has this changed?

File: src/Compiler/Facilities/DiagnosticsLogger.fsi [hunk 9]
```diff
@@ -466,15 +464,23 @@ type StackGuard =
 type CompilationGlobalsScope =
     new: diagnosticsLogger: DiagnosticsLogger * buildPhase: BuildPhase -> CompilationGlobalsScope
 
+    /// When dispoed, restores caller's diagnostics logger and build phase.
```
Comment (psfinaki) [inline src/Compiler/Facilities/DiagnosticsLogger.fsi:467]:
```suggestion
    /// When disposed, restores caller's diagnostics logger and build phase.
```

File: src/Compiler/Facilities/DiagnosticsLogger.fsi [hunk 10]
```diff
@@ -466,15 +464,23 @@ type StackGuard =
...
+
     interface IDisposable
 
     member DiagnosticsLogger: DiagnosticsLogger
 
     member BuildPhase: BuildPhase
 
-type CaptureDiagnosticsConcurrently =
-    new: unit -> CaptureDiagnosticsConcurrently
+module MultipleDiagnosticsLoggers =
```
Comment (psfinaki) [inline src/Compiler/Facilities/DiagnosticsLogger.fsi:476]:
```suggestion
module MultipleDiagnosticsLogger =
```
?

File: src/Compiler/Facilities/DiagnosticsLogger.fs [hunk 11]
```diff
@@ -920,16 +909,56 @@ type StackGuard(maxDepth: int, name: string) =
     static member GetDepthOption(name: string) =
         GetEnvInteger ("FSHARP_" + name + "StackGuardDepth") StackGuard.DefaultDepth
 
-type CaptureDiagnosticsConcurrently() =
-    let target = DiagnosticsThreadStatics.DiagnosticsLogger
-    let loggers = ResizeArray()
+module MultipleDiagnosticsLoggers =
```
Comment (psfinaki) [inline src/Compiler/Facilities/DiagnosticsLogger.fs:912]: This code looks a bit subtle, do you think we can add some explicit testing for this?

**Review by psfinaki — APPROVED** [review]
Okay so I finally gathered some mental resources for reviewing this :)

I added a few remarks and questions mostly for my understanding. Overall these are good changes since removing custom CEs reduces cognitive load of the code. Also I think it's a good moment to get this in.

Thanks!

File: src/Compiler/Utilities/illib.fs (hunk 7 above)
Comment (majocha) [reply src/Compiler/Utilities/illib.fs:172]: `NodeCode.RunImmediate` used to do this exception flattening, so probably to get it to 1:1 parity for it. At this moment I frankly don't remember if this was needed to get all the test passing.

File: src/Compiler/Driver/GraphChecking/GraphProcessing.fs (hunk 8 above)
Comment (majocha) [reply src/Compiler/Driver/GraphChecking/GraphProcessing.fs:184]: This probably should be restored. It's a remnant of fighting what looked to me like odd deadlocks, and [this blogpost](https://devblogs.microsoft.com/premier-developer/the-danger-of-taskcompletionsourcet-class/) inspired me to try it.

File: src/Compiler/Facilities/DiagnosticsLogger.fsi (hunk 10 above)
Comment (majocha) [reply src/Compiler/Facilities/DiagnosticsLogger.fsi:476]: Yeah, naming is hard. Maybe this could be just `module DiagnosticsLogger`? So we could say `DiagnosticsLogger.Parallel`?
But the name should convey we really do `Async.Parallel` but handling multithreaded diagnostics properly.

File: src/Compiler/Facilities/DiagnosticsLogger.fs (hunk 11 above)
Comment (majocha) [reply src/Compiler/Facilities/DiagnosticsLogger.fs:912]: Right. Specifically as for now there is not test checking that the diags are committed in the right order and timely.

File: src/Compiler/Facilities/DiagnosticsLogger.fsi (hunk 10 above)
Comment (psfinaki) [reply src/Compiler/Facilities/DiagnosticsLogger.fsi:476]: Yeah why not just `DiagnosticsLogger` actually.

IMO if there are ambitions to do something in parallel (and name is an ambition in that regard) then the handling of multiple threads should be assumed - although this definitely very often mismatches reality :D

File: src/Compiler/Utilities/illib.fs (hunk 7 above)
Comment (psfinaki) [reply src/Compiler/Utilities/illib.fs:172]: Not a big deal, but I noticed a dedicated "FlattenException" function in the PR, maybe it can be reused here as well?

Comment (majocha) [reply src/Compiler/Utilities/illib.fs:172]: Yeah there is also such utility function in FSharp.Editor. There is room for some code clean-up. I'm just afraid if this PR stops passing CI I'll never get around to fixing it again :D

File: src/Compiler/Facilities/DiagnosticsLogger.fsi (hunk 10 above)
Comment (majocha) [reply src/Compiler/Facilities/DiagnosticsLogger.fsi:476]: Now I remembered, there is already:
https://github.com/dotnet/fsharp/blob/0115c93d599a9f9eaec2f974c66c1944231a4913/src/Compiler/Driver/ParseAndCheckInputs.fs#L800

The name `MultipleDiagnosticsLoggers` here indicates the possibility of unifying this code.

Comment (psfinaki) [reply src/Compiler/Facilities/DiagnosticsLogger.fsi:476]: Oh okay. Maybe then just add a comment about this then.

File: src/Compiler/Utilities/illib.fs (hunk 7 above)
Comment (psfinaki) [reply src/Compiler/Utilities/illib.fs:172]: As you wish :) If this starts causing mysterious problems, we can probably keep the current version. But I'd still try to just remove it to see if it still causes some test failures.

File: src/Compiler/Facilities/DiagnosticsLogger.fs (hunk 11 above)
Comment (majocha) [reply src/Compiler/Facilities/DiagnosticsLogger.fs:912]: added a test.

File: src/Compiler/Utilities/illib.fs (hunk 7 above)
Comment (majocha) [reply src/Compiler/Utilities/illib.fs:172]: I'll try removing it in a separate commit.

File: src/Compiler/Facilities/DiagnosticsLogger.fs (hunk 11 above)
Comment (psfinaki) [reply src/Compiler/Facilities/DiagnosticsLogger.fs:912]: Thansk!

File: src/Compiler/Utilities/illib.fs (hunk 7 above)
Comment (psfinaki) [reply src/Compiler/Utilities/illib.fs:172]: Sure!

Comment (majocha) [reply src/Compiler/Utilities/illib.fs:172]: OK, I did some investigation about this and refreshed my memory. Turns out, apart from tests, the only occurrence of this particular `Async.RunImmediate` we have in `main` happens here in `StackGuard.Guard`

https://github.com/dotnet/fsharp/blob/0115c93d599a9f9eaec2f974c66c1944231a4913/src/Compiler/Facilities/DiagnosticsLogger.fs#L907

The rest are `NodeCode.RunImmediate` and `NodeCode.RunImmediateWithoutCancellation`, which, unlike the `Async` version, do the unwraping of AggregateException.

Since I replaced all of that with `Async.RunImmediate` I had to add the unwrapping. For example this keeps green some tests that expect `OperationCancelledException` and not the `AggregateException` that the raw `TaskCompletionSource` emits.

If this ever causes any problem in the future it would be easy to restore the old behavior in that single spot.

TBH I'm not even sure if the current behavior of `StackGuard` wrapping `AggregateException` around the thrown exceptions is  the desired one. For sure changing it does not trip any tests we have.


To conclude, I think this is OK as is.

Comment (psfinaki) [reply src/Compiler/Utilities/illib.fs:172]: Alright then, thanks for the explanation!

**Comment by vzarytovskii** [comment]
@majocha i saw reports from @ForNeVeR that some of the changes result in the deadlock

**Comment by majocha** [comment]
@vzarytovskii ok, that is something I remember seeing back in February, but couldn't really reproduce recently.

I would try restoring these two `SwitchToThreadPool()` calls:
https://github.com/dotnet/fsharp/pull/16779/commits/27dd98971c379cbc2f4b728ca33889bb40dc184e

 @ForNeVeR, do the deadlocks happen in FSI?

**Comment by majocha** [comment]
ah, ok I'm seeing the issue now.
